File: Example Code1/RadioHub/Utils/SerialMessage.cpp
/**
 * @file    SerialMessage.cpp
 *
 * @brief   Implementation of class SerialMessage
 *
 * @note    This example code is free software: you can redistribute it and/or modify it.
 *
 *          This program is provided by IMST GmbH on an "AS IS" basis without
 *          any warranties in the hope that it will be useful.
 */

#include "SerialMessage.h"
#include "CRC16.h"

SerialMessage::SerialMessage()
{

}

SerialMessage::SerialMessage( uint8_t sapID, uint8_t msgID )
{
    InitRequest( sapID, msgID );
}

uint8_t
SerialMessage::GetSapID() const
{
    if ( count() >= ( Header_Size ) )
    {
        return (uint8_t)at( SapID_Index );
    }
    return InvalidSapID;
}

uint8_t
SerialMessage::GetMsgID() const
{
    if ( count() >= ( Header_Size ) )
    {
        return (uint8_t)at( MsgID_Index );
    }
    return InvalidMsgID;
}

uint8_t
SerialMessage::GetResponseStatus() const
{
    if ( count() > ( Status_Index ) )
    {
        return (uint8_t)at( Status_Index );
    }
    return InvalidStatus;
}

int
SerialMessage::GetPayloadLength() const
{
    if ( count() >= ( Header_Size ) )
    {
        return ( count() - ( Header_Size ) );
    }
    return InvalidLength;
}

uint8_t
SerialMessage::GetU8( int index ) const
{
    if ( count() >= ( index + 1 ) )
    {
        return (uint8_t)at( index );
    }
    return 0;
}

uint16_t
SerialMessage::GetU16( int index ) const
{
    if ( count() >= ( index + 2 ) )
    {
        return (uint16_t)( (uint8_t)at( index ) <<  0 ) |
               (uint16_t)( (uint8_t)at( 1 + index ) <<  8 );
    }
    return 0;
}

uint32_t
SerialMessage::GetU32( int index ) const
{
    if ( count() >= ( index + 4 ) )
    {
        return (uint32_t)( (uint8_t)at( index ) <<  0 ) |
               (uint32_t)( (uint8_t)at( 1 + index ) <<  8 ) |
               (uint32_t)( (uint8_t)at( 2 + index ) << 16 ) |
               (uint32_t)( (uint8_t)at( 3 + index ) << 24 );
    }
    return 0;
}

uint64_t
SerialMessage::GetU64( int index ) const
{
    if ( count() >= ( index + 8 ) )
    {
        uint32_t lo = GetU32( index );
        uint32_t hi = GetU32( index + 4 );

        return (uint64_t)( ( (uint64_t)hi << 32 ) |  lo );
    }
    return 0;
}


QByteArray
SerialMessage::GetPayload( int index, int size ) const
{
    if ( size != -1 )
    {
        if ( count() >= ( index + size ) )
        {
            // return remaining payload
            return mid( index, size );
        }
    }
    else
    {
        // return remaining part of payload
        if ( count() > ( index ) )
        {
            // return remaining payload
            return mid( index, size );
        }
    }

    // return empty array
    return QByteArray();
}

void
SerialMessage::InitRequest( uint8_t sapID, uint8_t msgID )
{
    // clear buffer for init
    clear();

    // attach HCI header
    append( sapID );
    append( msgID );
}

int
SerialMessage::Append( uint8_t value )
{
    // LSB first
    append( (uint8_t)( value ) );

    // 1 byte appended
    return 1;
}

int
SerialMessage::Append( uint16_t value )
{
    // LSB first
    append( (uint8_t)( value ) );
    append( (uint8_t)( value >> 8 ) );

    // 2 bytes appended
    return 2;
}

int
SerialMessage::Append( uint32_t value )
{
    // LSB first
    append( (uint8_t)( value ) );
    append( (uint8_t)( value >> 8 ) );
    append( (uint8_t)( value >> 16 ) );
    append( (uint8_t)( value >> 24 ) );

    // 4 bytes appended
    return 4;
}

int
SerialMessage::Append( uint64_t value )
{
    // LSB first
    append( (uint8_t)( value ) );
    append( (uint8_t)( value >> 8 ) );
    append( (uint8_t)( value >> 16 ) );
    append( (uint8_t)( value >> 24 ) );

    append( (uint8_t)( value >> 32 ) );
    append( (uint8_t)( value >> 40 ) );
    append( (uint8_t)( value >> 48 ) );
    append( (uint8_t)( value >> 56 ) );

    // 8 bytes appended
    return 8;
}

int
SerialMessage::Append_CRC16()
{
    CRC16   crc16;

    // get reference to base class
    const QByteArray& data = *this;

    // append one's complement of crc
    return Append( (uint16_t)crc16.Calc_X25( data ) );
}

bool
SerialMessage::CheckCRC16() const
{
    CRC16   crc16;

    // get reference to base class
    const QByteArray& data = *this;

    return crc16.Check_X25( data );
}

void
SerialMessage::RemoveCRC16()
{
    if ( count() >= Min_Size )
    {
        // remove trailing crc bytes
        chop( CRC_Size );
    }
}

File: Example Code2.py/RadioHub/Utils/SerialMessage.py
'''
 @file    SerialMessage.py
 
 @brief   This class provides basic functions for encoding and decoding of HCI messages.
 
 @note    This example code is free software: you can redistribute it and/or modify it.
 
          This program is provided by IMST GmbH on an "AS IS" basis without
          any warranties in the hope that it will be useful.
 '''

from crccheck.crc import Crc16IsoHdlc
import struct
import time

class SerialMessage:

    Status_Index        =   0x00
    EventData_Index     =   0x00
    ResponseData_Index  =   0x01
      
    def __init__( self, sap_id = None, msg_id = None, payload = None ):
        self.sap_id = sap_id
        self.msg_id = msg_id

        if payload == None:
            self.payload = b''
        else:    
            self.payload = payload

        self.fcs = None

    def calc_crc( self ):
        data = bytes( [ self.sap_id ] ) + bytes( [ self.msg_id ] ) + self.payload
        crc = Crc16IsoHdlc.calc( data )
        self.fcs = struct.pack('<H', crc )

    def encode( self ):
        self.calc_crc()
        data = bytes( [ self.sap_id ] ) + bytes( [ self.msg_id ] ) + self.payload + bytes( self.fcs )
        return data
        
    def decode( self, data ) -> bool:

        if ( len( data ) < 4):
            return False

        self.sap_id = data[ 0 ]
        self.msg_id = data[ 1 ]
        self.payload = data [ 2: -2 ]
        self.calc_crc()

        return ( self.fcs == data[ -2: ] )
         
    def getPayloadLength( self ) -> int:      
        return len( self.payload )
    
    def getU8( self, index : int ) -> int:
        return int( self.payload[ index ] )
    
    def getI8( self, index : int ) -> int:
        i8 = struct.unpack('b', self.payload[ index:index+1 ])[0]
        return int( i8 )
    
    def getStatus( self ) -> int:
        return self.getU8( self.Status_Index )
        
    def getU16( self, index : int ) -> int:
        # little-endian
        u16 = struct.unpack('<H', self.payload[ index:index+2 ])[0]
        return int( u16 )

    def getU16_MSB( self, index : int ) -> int:
        # big-endian
        u16 = struct.unpack('>H', self.payload[ index:index+2 ])[0]
        return int( u16 )

    def getU32( self, index : int ) -> int:
        # litte-endian
        u32 = struct.unpack('<I', self.payload[ index:index+4 ])[0]
        return int( u32 )

    def getU32_MSB( self, index : int ) -> int:
        # big-endian
        u32 = struct.unpack('>I', self.payload[ index:index+4 ])[0]
        return int( u32 )

    def getString( self, index : int, size : int ) -> str:
        if ( size == -1 ):
            size = len( self.payload ) - index
        help = self.payload[ index:index+size]    
        return help.decode( 'utf-8' )

    def getHexString( self, index : int, size : int ) ->str:
        if ( size == -1 ):
            size = len( self.payload ) - index
        help = self.payload[ index:index+size]    
        s = help.hex()
        return '-'.join( s[i:i+2] for i in range( 0, len( s ), 2 ) )

    def getHexString_LSB( self, index : int, size : int ) ->str:
        if ( size == -1 ):
            size = len( self.payload ) - index
        help = self.payload[ index:index+size]    
        help2 = help[::-1]
        s = help2.hex()
        return '-'.join( s[i:i+2] for i in range( 0, len( s ), 2 ) )

    def getDateTime( self, index : int ):
        u32 = struct.unpack('<I', self.payload[ index:index+4 ])[0]
        return time.gmtime( self.getU32( index ) )
    
        
        